## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [ ]:
import seaborn as sns

titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
867,0,1,male,31.0,0,0,50.4958,S,First,man,True,A,Southampton,no,True
530,1,2,female,2.0,1,1,26.0000,S,Second,child,False,NaN,Southampton,yes,False
9,1,2,female,14.0,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False
461,0,3,male,34.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
735,0,3,male,28.5,0,0,16.1000,S,Third,man,True,NaN,Southampton,no,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [47]:
# ваш код
titanic_data.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [48]:
# ваш код
titanic_data = titanic_data.dropna(axis=1, thresh=len(titanic_data) * 0.5 + 1)
titanic_data = titanic_data.dropna(axis=0, thresh=titanic_data.shape[1] * 0.5 + 1)
titanic_data["age"].isnull().sum()

np.int64(177)

### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [ ]:
# ваш код
for elem in ["man", "woman", "child"]:
    median = round(titanic_data.loc[titanic_data["who"] == elem, "age"].median())
    titanic_data.loc[(titanic_data["age"].isnull()) & (titanic_data["who"] == elem), "age"] = median

### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [69]:
# ваш код
titanic_data = titanic_data.dropna(thresh=titanic_data.shape[1] - 1)
titanic_data.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [51]:
# ваш код
titanic_data["embark_town"].value_counts().idxmax()

'Southampton'

### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [52]:
# ваш код
round(titanic_data["survived"].mean() * 100, 2)

np.float64(38.25)

### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [53]:
# ваш код
titanic_data[titanic_data["survived"] == 1]["embark_town"].value_counts()

embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: count, dtype: int64

### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [67]:
# ваш код
survived_by_class = titanic_data[titanic_data["survived"] == 1]["class"].value_counts()
total_by_class = titanic_data["class"].value_counts()
(survived_by_class / total_by_class * 100).round(2)

class
First     62.62
Second    47.28
Third     24.24
Name: count, dtype: float64

### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [59]:
# ваш код
rich_passengers = titanic_data[titanic_data["fare"] > 100]
round(rich_passengers["survived"].sum() / len(rich_passengers) * 100, 2)

np.float64(73.58)

### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [ ]:
# ваш код
((titanic_data["who"] == "child") & (titanic_data["alone"])).sum()

np.int64(6)

Какие выводы вы можете сделать о выживших пассажирах Титаника? 